src/backend/lib/upgrade-manager.js

```diff
@@ -108,6 +114,8 @@ class UpgradeManager {
     const result = validate(this.state);
     if (result) {
       const msg = result.join(",");
+      // Validation error: this indicates an internal error that should never happen.
+      // If it does: definitely a fatal error.
```

Comment:
Ok so I think this should be considered an error state then. The UpgradeManager is in an error state once this happens. However, I feel like we are doing runtime validation of something that should be a test or a type check. Does any input of `this.state` come from userland (e.g. read from somewhere external) or is state always generated by the code here, e.g. if this doesn't validate is it because there is a bug in this code, or could it be bad data getting into here somehow?

Reply:
Each component (`Server, Search, Download, Check`) emits its own `{state, context}`, which parent components (`UpgradeDownload, UpgradeManager`) compose into a larger state object.

Reply:
> if this doesn't validate is it because there is a bug in this code, or could it be bad data getting into here somehow?

Only code bugs should be able to cause an error here.